File: web/src/lib/brand/brand.test.ts
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';

import {
  BRAND_FILE_NAME,
  BrandError,
  assertBrand,
  brandCustomProperties,
  brandStyleSheet,
  parseBrand,
  type Brand,
} from './index';

// The fixture brand the Go package is tested against, read from where
// the Go package keeps it. Not a copy: the same bytes, so a change to
// the file that only one of the two readers can cope with fails here.
//
// It is deliberately unlike the brand this repository ships — different
// name, palette, currency, default language and place (ADR-0004) — so a
// test that passes is evidence the code reads its configuration rather
// than remembering the product it was written for.
const fixtureUrl = new URL(
  `../../../../internal/platform/brand/testdata/fixture/${BRAND_FILE_NAME}`,
  import.meta.url,
);
const fixtureJson = readFileSync(fileURLToPath(fixtureUrl), 'utf8');

function fixture(): Brand {
  return parseBrand(fixtureJson);
}

/** The fixture brand as a plain object, for building invalid variants. */
function fixtureObject(): Record<string, unknown> {
  return JSON.parse(fixtureJson) as Record<string, unknown>;
}

describe('parseBrand', () => {
  it('reads the same fixture file the Go loader reads, to the same values', () => {
    expect(fixture()).toEqual({
      id: 'zephyra',
      name: 'Zephyra',
      legal: {
        entity: 'Zephyra Fixture Kooperativ AB',
        jurisdiction: 'SE',
        address: 'Kungsgatan 1, 411 19 Göteborg, Sweden',
        documents: {
          terms: { id: 'zephyra-terms', version: '3.1.0' },
          privacy: { id: 'zephyra-privacy', version: '2026-05-04' },
          cookies: { id: 'zephyra-cookies', version: '1.0.0' },
        },
      },
      domains: {
        primary: 'zephyra.example',
        aliases: ['www.zephyra.example', 'zephyra-forna.example'],
      },
      support: {
        general: 'hej@zephyra.example',
        legal: 'juridik@zephyra.example',
        privacy: 'dataskydd@www.zephyra.example',
      },
      assets: {
        logo: '/brand/zephyra/logo.svg',
        logoDark: '/brand/zephyra/logo-dark.svg',
        favicon: '/brand/zephyra/favicon.ico',
      },
      theme: {
        colours: {
          bg: '#0b1d26',
          surface: '#123243',
          text: '#eef6f8',
          accent: '#2ec4b6',
          'accent-2': '#ffbf47',
        },
        typography: {
          heading: '"Fraunces", Georgia, serif',
          body: '"Public Sans", system-ui, sans-serif',
          headingWeight: 700,
        },
      },
      defaults: { language: 'sv', place: 'goteborg', currency: 'SEK' },
      payout: { descriptor: 'ZEPHYRA CASHBACK' },
      features: {
        cashback: { wallet: true, withdrawal: true },
        news: { reader: false },
      },
    });
  });

  it('refuses anything that is not JSON', () => {
    expect(() => parseBrand('not json')).toThrow(BrandError);
    expect(() => parseBrand('not json')).toThrow(/not valid JSON/);
  });
});

describe('assertBrand', () => {
  it('accepts the fixture brand', () => {
    expect(() => {
      assertBrand(fixtureObject());
    }).not.toThrow();
  });

  const cases: ReadonlyArray<{
    readonly name: string;
    readonly mutate: (brand: Record<string, unknown>) => void;
    readonly problem: RegExp;
  }> = [
    {
      name: 'a missing field',
      mutate: (brand) => {
        delete brand['payout'];
      },
      problem: /payout is missing/,
    },
    {
      name: 'a missing nested field',
      mutate: (brand) => {
        delete (brand['legal'] as Record<string, unknown>)['entity'];
      },
      problem: /legal\.entity is missing/,
    },
    {
      name: 'a field of the wrong primitive type',
      mutate: (brand) => {
        brand['name'] = 42;
      },
      problem: /name is a number, want string/,
    },
    {
      name: 'a number where the schema wants one and a string arrived',
      mutate: (brand) => {
        (brand['theme'] as { typography: Record<string, unknown> }).typography['headingWeight'] = '700';
      },
      problem: /headingWeight is a string, want number/,
    },
    {
      name: 'a boolean flag that is not a boolean',
      mutate: (brand) => {
        (brand['features'] as { cashback: Record<string, unknown> }).cashback['wallet'] = 'yes';
      },
      problem: /features\.cashback\.wallet is a string, want boolean/,
    },
    {
      name: 'an object where a list belongs',
      mutate: (brand) => {
        (brand['domains'] as Record<string, unknown>)['aliases'] = { first: 'x' };
      },
      problem: /aliases is an object, want a list/,
    },
    {
      name: 'a list element of the wrong type',
      mutate: (brand) => {
        (brand['domains'] as Record<string, unknown>)['aliases'] = ['ok', 7];
      },
      problem: /aliases\[1\] is a number, want string/,
    },
    {
      name: 'a list where a map belongs',
      mutate: (brand) => {
        (brand['theme'] as Record<string, unknown>)['colours'] = ['#000000'];
      },
      problem: /colours is a list, want an object/,
    },
    {
      name: 'a nested interface that is not an object',
      mutate: (brand) => {
        brand['legal'] = 'Zephyra Fixture Kooperativ AB';
      },
      problem: /legal is a string, want an object/,
    },
    {
      name: 'a null where an interface belongs',
      mutate: (brand) => {
        brand['payout'] = null;
      },
      problem: /payout is null, want an object/,
    },
    {
      // A key nobody reads is a value that silently does not apply,
      // which is how a surface keeps the previous brand's colour.
      name: 'a key the schema has never heard of',
      mutate: (brand) => {
        brand['tagline'] = 'unknown to the schema';
      },
      problem: /tagline is not part of the brand schema/,
    },
    {
      name: 'a nested key the schema has never heard of',
      mutate: (brand) => {
        (brand['payout'] as Record<string, unknown>)['statementPrefix'] = 'ZEP';
      },
      problem: /payout\.statementPrefix is not part of the brand schema/,
    },
    {
      name: 'a map entry of the wrong shape',
      mutate: (brand) => {
        (brand['legal'] as { documents: Record<string, unknown> }).documents['terms'] = 'v3';
      },
      problem: /legal\.documents\.terms is a string, want an object/,
    },
  ];

  for (const testCase of cases) {
    it(`refuses ${testCase.name}`, () => {
      const brand = fixtureObject();
      testCase.mutate(brand);
      expect(() => {
        assertBrand(brand);
      }).toThrow(BrandError);
      expect(() => {
        assertBrand(brand);
      }).toThrow(testCase.problem);
    });
  }

  it('refuses a brand that is not an object at all', () => {
    expect(() => {
      assertBrand([]);
    }).toThrow(/the brand is a list, want an object/);
    expect(() => {
      assertBrand(null);
    }).toThrow(/the brand is null, want an object/);
  });

  it('names every problem at once, in a stable order', () => {
    const brand = fixtureObject();
    delete brand['payout'];
    brand['name'] = 42;
    let message = '';
    try {
      assertBrand(brand);
    } catch (error) {
      message = (error as BrandError).message;
    }
    expect(message).toMatch(/name is a number/);
    expect(message).toMatch(/payout is missing/);
    expect(message.indexOf('name is')).toBeLessThan(message.indexOf('payout is'));
  });

  // `in` walks the prototype chain. These two cases are what that costs,
  // and they are reachable because this module validates values built in
  // code, not only objects that came out of JSON.parse.
  it('does not read an inherited property as a field that is present', () => {
    const complete = fixtureObject();
    const payout = complete['payout'];
    delete complete['payout'];

    const inherited = Object.create({ payout }) as Record<string, unknown>;
    for (const [key, value] of Object.entries(complete)) {
      inherited[key] = value;
    }

    expect(() => {
      assertBrand(inherited);
    }).toThrow(/payout is missing/);
  });

  it('flags a key that only looks like a schema member because everything inherits it', () => {
    for (const key of ['toString', 'constructor', 'valueOf']) {
      const brand = fixtureObject();
      brand[key] = 'not part of the schema';
      expect(() => {
        assertBrand(brand);
      }).toThrow(new RegExp(`${key} is not part of the brand schema`));
    }
  });

  it('accepts a brand with no prototype at all', () => {
    const bare = Object.create(null) as Record<string, unknown>;
    for (const [key, value] of Object.entries(fixtureObject())) {
      bare[key] = value;
    }
    expect(() => {
      assertBrand(bare);
    }).not.toThrow();
  });

  // JSON carries no NaN, so this can only arrive from a value built in
  // code — which the loader is also used for.
  it('refuses a number that is not finite', () => {
    const brand = fixtureObject();
    (brand['theme'] as { typography: Record<string, unknown> }).typography['headingWeight'] = Number.NaN;
    expect(() => {
      assertBrand(brand);
    }).toThrow(/headingWeight is not a finite number/);
  });
});

describe('brandCustomProperties', () => {
  it('fills the variables the design system already declares', () => {
    expect(brandCustomProperties(fixture())).toEqual({
      '--color-accent': '#2ec4b6',
      '--color-accent-2': '#ffbf47',
      '--color-bg': '#0b1d26',
      '--color-surface': '#123243',
      '--color-text': '#eef6f8',
      '--font-heading': '"Fraunces", Georgia, serif',
      '--font-heading-weight': '700',
      '--font-body': '"Public Sans", system-ui, sans-serif',
    });
  });

  it('emits colour tokens in name order, so two builds of one brand match', () => {
    const names = Object.keys(brandCustomProperties(fixture())).filter((name) => name.startsWith('--color-'));
    expect(names).toEqual([...names].sort());
  });

  it('carries a brand token the design system has never seen', () => {
    const brand = fixtureObject();
    (brand['theme'] as { colours: Record<string, string> }).colours['highlight'] = '#ff00ff';
    assertBrand(brand);
    expect(brandCustomProperties(brand)['--color-highlight']).toBe('#ff00ff');
  });

  for (const [name, value] of [
    ['a declaration terminator', '#000000; background: url(x)'],
    ['a closing brace', '#000000 }'],
    ['a tag', '#000000</style'],
    ['a newline', '#000000\n'],
  ] as ReadonlyArray<readonly [string, string]>) {
    it(`refuses a value carrying ${name}`, () => {
      const brand = fixtureObject();
      (brand['theme'] as { colours: Record<string, string> }).colours['bg'] = value;
      assertBrand(brand);
      expect(() => brandCustomProperties(brand)).toThrow(BrandError);
      expect(() => brandCustomProperties(brand)).toThrow(/break out of the stylesheet/);
    });
  }

  // The schema promises Record<string, string>, and a map's KEYS carry
  // no type at all — so nothing upstream can vouch for a token name, and
  // the name is what goes into `--color-${token}`.
  for (const token of ['bg}', 'bg<style', 'Accent Two', '', '-leading-hyphen', 'bg;color:red']) {
    it(`refuses the colour token name ${JSON.stringify(token)}`, () => {
      const brand = fixtureObject();
      (brand['theme'] as { colours: Record<string, string> }).colours[token] = '#ffbf47';
      assertBrand(brand);
      expect(() => brandCustomProperties(brand)).toThrow(BrandError);
      expect(() => brandCustomProperties(brand)).toThrow(/is not a CSS token name/);
    });
  }

  it('refuses a font stack that would break out of the stylesheet', () => {
    const brand = fixtureObject();
    (brand['theme'] as { typography: Record<string, unknown> }).typography['body'] = 'serif; }';
    assertBrand(brand);
    expect(() => brandCustomProperties(brand)).toThrow(/theme\.typography\.body/);

    const heading = fixtureObject();
    (heading['theme'] as { typography: Record<string, unknown> }).typography['heading'] = 'serif; }';
    assertBrand(heading);
    expect(() => brandCustomProperties(heading)).toThrow(/theme\.typography\.heading/);
  });
});

describe('brandStyleSheet', () => {
  it('writes the custom properties as a rule', () => {
    expect(brandStyleSheet(fixture())).toBe(
      [
        ':root {',
        '  --color-accent: #2ec4b6;',
        '  --color-accent-2: #ffbf47;',
        '  --color-bg: #0b1d26;',
        '  --color-surface: #123243;',
        '  --color-text: #eef6f8;',
        '  --font-heading: "Fraunces", Georgia, serif;',
        '  --font-heading-weight: 700;',
        '  --font-body: "Public Sans", system-ui, sans-serif;',
        '}',
        '',
      ].join('\n'),
    );
  });

  it('accepts a selector other than the document root', () => {
    expect(brandStyleSheet(fixture(), '.brand-preview')).toMatch(/^\.brand-preview \{\n/);
    expect(brandStyleSheet(fixture(), '[data-brand="zephyra"] .card')).toMatch(/^\[data-brand="zephyra"\] \.card \{\n/);
  });

  // Every caller today passes a literal, but this is exported as a
  // general utility and its argument is an input like any other.
  for (const selector of [':root} body { display: none', ':root</style', ':root;', ':root\n}', '', '   ']) {
    it(`refuses the selector ${JSON.stringify(selector)}`, () => {
      expect(() => brandStyleSheet(fixture(), selector)).toThrow(BrandError);
      expect(() => brandStyleSheet(fixture(), selector)).toThrow(/is not a usable stylesheet selector/);
    });
  }
});

File: internal/platform/brand/brand_test.go
package brand_test

import (
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"

	"github.com/Nomos-N4s/apivo-news/internal/platform/brand"
)

// fixtureDir holds the fixture brand ADR-0004 requires: a complete,
// valid brand that is deliberately unlike the one this repository
// currently ships — different name, different palette, different
// currency, different default language and place. Every test here works
// against it rather than against the live brand, so a test that passes
// is evidence the code reads the configuration rather than remembering
// the product it was written for.
const fixtureDir = "testdata/fixture"

// fixture is what testdata/fixture/brand.json means, written out
// longhand. The duplication is the assertion: if the file and this value
// ever disagree, one of them changed without anyone deciding to.
func fixture() brand.Brand {
	return brand.Brand{
		ID:   "zephyra",
		Name: "Zephyra",
		Legal: brand.Legal{
			Entity:       "Zephyra Fixture Kooperativ AB",
			Jurisdiction: "SE",
			Address:      "Kungsgatan 1, 411 19 Göteborg, Sweden",
			Documents: map[string]brand.Document{
				"terms":   {ID: "zephyra-terms", Version: "3.1.0"},
				"privacy": {ID: "zephyra-privacy", Version: "2026-05-04"},
				"cookies": {ID: "zephyra-cookies", Version: "1.0.0"},
			},
		},
		Domains: brand.Domains{
			Primary: "zephyra.example",
			Aliases: []string{"www.zephyra.example", "zephyra-forna.example"},
		},
		Support: brand.Support{
			General: "hej@zephyra.example",
			Legal:   "juridik@zephyra.example",
			Privacy: "dataskydd@www.zephyra.example",
		},
		Assets: brand.Assets{
			Logo:     "/brand/zephyra/logo.svg",
			LogoDark: "/brand/zephyra/logo-dark.svg",
			Favicon:  "/brand/zephyra/favicon.ico",
		},
		Theme: brand.Theme{
			Colours: map[string]string{
				"bg":       "#0b1d26",
				"surface":  "#123243",
				"text":     "#eef6f8",
				"accent":   "#2ec4b6",
				"accent-2": "#ffbf47",
			},
			Typography: brand.Typography{
				Heading:       `"Fraunces", Georgia, serif`,
				Body:          `"Public Sans", system-ui, sans-serif`,
				HeadingWeight: 700,
			},
		},
		Defaults: brand.Defaults{Language: "sv", Place: "goteborg", Currency: "SEK"},
		Payout:   brand.Payout{Descriptor: "ZEPHYRA CASHBACK"},
		Features: map[string]map[string]bool{
			"cashback": {"wallet": true, "withdrawal": true},
			"news":     {"reader": false},
		},
	}
}

func TestLoadDirReadsTheFixtureBrand(t *testing.T) {
	t.Parallel()

	got, err := brand.LoadDir(fixtureDir)
	if err != nil {
		t.Fatalf("LoadDir(%q): %v", fixtureDir, err)
	}
	if want := fixture(); !reflect.DeepEqual(got, want) {
		t.Errorf("LoadDir(%q)\n got %+v\nwant %+v", fixtureDir, got, want)
	}
}

func TestLoadReadsAnyFileSystem(t *testing.T) {
	t.Parallel()

	got, err := brand.Load(os.DirFS(fixtureDir), brand.FileName)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	if want := fixture(); !reflect.DeepEqual(got, want) {
		t.Errorf("Load\n got %+v\nwant %+v", got, want)
	}
}

func TestLoadReportsAMissingFile(t *testing.T) {
	t.Parallel()

	_, err := brand.LoadDir(filepath.Join("testdata", "no-such-brand"))
	if err == nil {
		t.Fatal("LoadDir on a missing directory returned no error")
	}
	if !errors.Is(err, os.ErrNotExist) {
		t.Errorf("error %v does not wrap os.ErrNotExist", err)
	}
}

func TestParseRejectsMalformedDefinitions(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name string
		data string
		want string
	}{
		{
			name: "not json at all",
			data: "not json",
			want: "brand: decode",
		},
		{
			// A key nobody reads is a value that silently does not
			// apply, which is how a surface keeps the previous
			// brand's colour and nobody can say why.
			name: "unknown field",
			data: `{"id":"x","tagline":"unknown to the schema"}`,
			want: "unknown field",
		},
		{
			name: "wrong type for a field",
			data: `{"id":42}`,
			want: "brand: decode",
		},
	}

	for _, testCase := range tests {
		t.Run(testCase.name, func(t *testing.T) {
			t.Parallel()

			_, err := brand.Parse([]byte(testCase.data))
			if err == nil {
				t.Fatal("Parse accepted a malformed definition")
			}
			if !strings.Contains(err.Error(), testCase.want) {
				t.Errorf("error %q does not mention %q", err, testCase.want)
			}
		})
	}
}

func TestParseRejectsAWellFormedFileThatBreaksTheRules(t *testing.T) {
	t.Parallel()

	// Valid JSON, valid field names, and a brand no surface could be
	// rendered from. Decoding is not acceptance.
	_, err := brand.Parse([]byte(`{"id":"zephyra","name":"Zephyra"}`))
	if err == nil {
		t.Fatal("Parse accepted a brand with nothing but a name")
	}
	if !errors.Is(err, brand.ErrInvalid) {
		t.Errorf("error %v does not wrap ErrInvalid", err)
	}
}

func TestParseAcceptsTheFixtureFile(t *testing.T) {
	t.Parallel()

	data, err := os.ReadFile(filepath.Join(fixtureDir, brand.FileName))
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	if _, err := brand.Parse(data); err != nil {
		t.Fatalf("Parse rejected the fixture brand: %v", err)
	}
}

func TestValidateAcceptsTheFixtureBrand(t *testing.T) {
	t.Parallel()

	if err := fixture().Validate(); err != nil {
		t.Fatalf("the fixture brand is invalid: %v", err)
	}
}

func TestValidateNamesEveryBrokenRule(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name   string
		mutate func(*brand.Brand)
		want   string
	}{
		{
			name:   "id is not a slug",
			mutate: func(b *brand.Brand) { b.ID = "Zephyra Brand" },
			want:   "id \"Zephyra Brand\" is not a slug",
		},
		{
			name:   "name is empty",
			mutate: func(b *brand.Brand) { b.Name = "  " },
			want:   "name is empty",
		},
		{
			name:   "name is padded",
			mutate: func(b *brand.Brand) { b.Name = "Zephyra " },
			want:   "leading or trailing whitespace",
		},
		{
			name:   "legal entity is empty",
			mutate: func(b *brand.Brand) { b.Legal.Entity = "" },
			want:   "legal.entity is empty",
		},
		{
			name:   "jurisdiction is not a country code",
			mutate: func(b *brand.Brand) { b.Legal.Jurisdiction = "sweden" },
			want:   "legal.jurisdiction \"sweden\"",
		},
		{
			name:   "legal address is empty",
			mutate: func(b *brand.Brand) { b.Legal.Address = "" },
			want:   "legal.address is empty",
		},
		{
			name:   "terms are missing",
			mutate: func(b *brand.Brand) { delete(b.Legal.Documents, brand.DocumentTerms) },
			want:   "legal.documents is missing \"terms\"",
		},
		{
			name:   "privacy notice is missing",
			mutate: func(b *brand.Brand) { delete(b.Legal.Documents, brand.DocumentPrivacy) },
			want:   "legal.documents is missing \"privacy\"",
		},
		{
			name: "document has no identifier",
			mutate: func(b *brand.Brand) {
				b.Legal.Documents[brand.DocumentTerms] = brand.Document{Version: "1.0.0"}
			},
			want: "legal.documents[\"terms\"].id is empty",
		},
		{
			name: "document has no version",
			mutate: func(b *brand.Brand) {
				b.Legal.Documents[brand.DocumentTerms] = brand.Document{ID: "zephyra-terms"}
			},
			want: "legal.documents[\"terms\"].version is empty",
		},
		{
			name:   "primary domain carries a scheme",
			mutate: func(b *brand.Brand) { b.Domains.Primary = "https://zephyra.example" },
			want:   "domains.primary",
		},
		{
			name:   "alias is not a host",
			mutate: func(b *brand.Brand) { b.Domains.Aliases = []string{"Zephyra.Example/path"} },
			want:   "domains.aliases[0]",
		},
		{
			name: "alias repeats the primary domain",
			mutate: func(b *brand.Brand) {
				b.Domains.Aliases = []string{"zephyra.example"}
			},
			want: "is a duplicate",
		},
		{
			name:   "support address is not an address",
			mutate: func(b *brand.Brand) { b.Support.General = "hej at zephyra example" },
			want:   "support.general",
		},
		{
			// The rebrand that looks finished and is not: the name
			// changed, the mailbox did not.
			name:   "support address is left on another brand's domain",
			mutate: func(b *brand.Brand) { b.Support.Legal = "juridik@someone-else.example" },
			want:   "is not on one of the brand's own domains",
		},
		{
			name:   "asset path is an absolute url",
			mutate: func(b *brand.Brand) { b.Assets.Logo = "https://cdn.example/logo.svg" },
			want:   "assets.logo",
		},
		{
			name:   "asset path escapes the web root",
			mutate: func(b *brand.Brand) { b.Assets.Favicon = "/brand/../../etc/favicon.ico" },
			want:   "escapes the web root",
		},
		{
			name:   "a required colour token is missing",
			mutate: func(b *brand.Brand) { delete(b.Theme.Colours, brand.ColourAccent) },
			want:   "theme.colours is missing \"accent\"",
		},
		{
			name:   "colour token name is not a slug",
			mutate: func(b *brand.Brand) { b.Theme.Colours["Accent Two"] = "#ffbf47" },
			want:   "theme.colours key \"Accent Two\"",
		},
		{
			name:   "colour is not lower-case six-digit hex",
			mutate: func(b *brand.Brand) { b.Theme.Colours[brand.ColourText] = "#EEF" },
			want:   "is not a lower-case six-digit hex colour",
		},
		{
			name:   "heading font stack is empty",
			mutate: func(b *brand.Brand) { b.Theme.Typography.Heading = "" },
			want:   "theme.typography.heading is empty",
		},
		{
			name:   "body font stack is empty",
			mutate: func(b *brand.Brand) { b.Theme.Typography.Body = "" },
			want:   "theme.typography.body is empty",
		},
		{
			name:   "heading weight is outside the css range",
			mutate: func(b *brand.Brand) { b.Theme.Typography.HeadingWeight = 0 },
			want:   "headingWeight 0",
		},
		{
			// A region subtag keys no catalogue: the catalogues are
			// keyed by the primary subtag alone (constitution VII).
			name:   "default language carries a region",
			mutate: func(b *brand.Brand) { b.Defaults.Language = "sv-SE" },
			want:   "defaults.language \"sv-SE\"",
		},
		{
			name:   "default place is not a slug",
			mutate: func(b *brand.Brand) { b.Defaults.Place = "Göteborg" },
			want:   "defaults.place",
		},
		{
			name:   "default currency is not iso 4217",
			mutate: func(b *brand.Brand) { b.Defaults.Currency = "kr" },
			want:   "defaults.currency \"kr\"",
		},
		{
			name:   "payout descriptor is empty",
			mutate: func(b *brand.Brand) { b.Payout.Descriptor = "" },
			want:   "payout.descriptor is empty",
		},
		{
			name:   "payout descriptor would be truncated",
			mutate: func(b *brand.Brand) { b.Payout.Descriptor = "ZEPHYRA CASHBACK REWARDS" },
			want:   "longer than 22 characters",
		},
		{
			name:   "payout descriptor is padded",
			mutate: func(b *brand.Brand) { b.Payout.Descriptor = " ZEPHYRA" },
			want:   "leading or trailing whitespace",
		},
		{
			name:   "payout descriptor is not ascii",
			mutate: func(b *brand.Brand) { b.Payout.Descriptor = "ZEPHYRA ÅTERBÄRING" },
			want:   "card schemes will transliterate",
		},
		{
			name:   "no product declares any flags",
			mutate: func(b *brand.Brand) { b.Features = nil },
			want:   "features declares no product",
		},
		{
			name:   "product key is not a slug",
			mutate: func(b *brand.Brand) { b.Features["Cash Back"] = map[string]bool{"wallet": true} },
			want:   "features key \"Cash Back\"",
		},
		{
			name:   "flag key is not a slug",
			mutate: func(b *brand.Brand) { b.Features["cashback"]["Instant Payout"] = true },
			want:   "features[\"cashback\"] key \"Instant Payout\"",
		},
	}

	for _, testCase := range tests {
		t.Run(testCase.name, func(t *testing.T) {
			t.Parallel()

			b := fixture()
			testCase.mutate(&b)

			err := b.Validate()
			if err == nil {
				t.Fatal("Validate accepted a brand it should have rejected")
			}
			if !errors.Is(err, brand.ErrInvalid) {
				t.Errorf("error %v does not wrap ErrInvalid", err)
			}
			if !strings.Contains(err.Error(), testCase.want) {
				t.Errorf("error %q does not mention %q", err, testCase.want)
			}
		})
	}
}

func TestValidateReportsEveryProblemAtOnce(t *testing.T) {
	t.Parallel()

	// An empty brand breaks nearly every rule. Reporting them one per
	// run would make filling in a brand file an afternoon of guessing.
	err := brand.Brand{}.Validate()
	if err == nil {
		t.Fatal("the zero brand validated")
	}
	for _, want := range []string{"id \"\"", "name is empty", "legal.entity is empty", "payout.descriptor is empty"} {
		if !strings.Contains(err.Error(), want) {
			t.Errorf("error %q does not mention %q", err, want)
		}
	}
}

func TestFeatureIsOffUnlessDeclaredOn(t *testing.T) {
	t.Parallel()

	b := fixture()
	tests := []struct {
		product string
		flag    string
		want    bool
	}{
		{product: "cashback", flag: "wallet", want: true},
		{product: "cashback", flag: "withdrawal", want: true},
		{product: "news", flag: "reader", want: false},
		{product: "cashback", flag: "instant-payout", want: false},
		{product: "grocery", flag: "wallet", want: false},
	}

	for _, testCase := range tests {
		t.Run(testCase.product+"/"+testCase.flag, func(t *testing.T) {
			t.Parallel()

			if got := b.Feature(testCase.product, testCase.flag); got != testCase.want {
				t.Errorf("Feature(%q, %q) = %t, want %t", testCase.product, testCase.flag, got, testCase.want)
			}
		})
	}
}

func TestDocumentReportsWhetherTheBrandHasOne(t *testing.T) {
	t.Parallel()

	b := fixture()

	doc, ok := b.Document(brand.DocumentTerms)
	if !ok {
		t.Fatal("the fixture brand has no terms document")
	}
	if doc.Version != "3.1.0" {
		t.Errorf("terms version = %q, want %q", doc.Version, "3.1.0")
	}

	if _, ok := b.Document("shareholder-agreement"); ok {
		t.Error("Document reported a kind the brand does not publish")
	}
}

func TestHostsAndAddressesAndPathsKeepTheirOrder(t *testing.T) {
	t.Parallel()

	b := fixture()

	wantHosts := []string{"zephyra.example", "www.zephyra.example", "zephyra-forna.example"}
	if got := b.Domains.Hosts(); !reflect.DeepEqual(got, wantHosts) {
		t.Errorf("Hosts() = %v, want %v", got, wantHosts)
	}

	wantAddresses := []string{"hej@zephyra.example", "juridik@zephyra.example", "dataskydd@www.zephyra.example"}
	if got := b.Support.Addresses(); !reflect.DeepEqual(got, wantAddresses) {
		t.Errorf("Addresses() = %v, want %v", got, wantAddresses)
	}

	wantPaths := []string{"/brand/zephyra/logo.svg", "/brand/zephyra/logo-dark.svg", "/brand/zephyra/favicon.ico"}
	if got := b.Assets.Paths(); !reflect.DeepEqual(got, wantPaths) {
		t.Errorf("Paths() = %v, want %v", got, wantPaths)
	}
}
